modules/tensor_mechanics/include/utils/StabilizationUtils.h

```diff
@@ -0,0 +1,31 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "MooseTypes.h"
+
+namespace StabilizationUtils
+{
+template <typename Functor>
+auto
+elementAverage(const Functor & f, const MooseArray<Real> & JxW, const MooseArray<Real> & coord)
+    -> decltype(f(std::declval<unsigned int>()))
```

Comment:
the training return type is not needed in C++17 anymore

Reply:
wait, what did you want to do here? wouldn't that version return a reference to a temporary?

Reply:
> the training return type is not needed in C++17 anymore

Yeah, this reminds me that this PR started back when we didn't have c++17 support yet :-)

Reply:
> wait, what did you want to do here? wouldn't that version return a reference to a temporary?

I am not sure. I just checked the documentation on decltype: https://en.cppreference.com/w/cpp/language/decltype
I think I am assuming `f(std::declval<unsigned int>())` to return a prvalue, and according to the doc the `decltype` should return `T`, not `T &`.

Reply:
Oh, right, it shoud return whatever type `f(...)` returns. But it seems to me you could have just written `decltype(f(0))` here, or not? Anypoop, leaving out that stuff makes it a lot less intimidating :-D